### PR TITLE
decompogen: supporting lambda function

### PR DESCRIPTION
**Pull request by Shekharrajak**
In this branch

```
    >>> decompogen(Lambda(x, sin(cos(x))), x)
    [sin(x), cos(x)] 
    >>>> decompogen(lambda x : sin(cos(x)), x)
    [sin(x), cos(x), x]

```

### CODE REVIEW

**Comment by Shekharrajak** [comment]
Still I finding why there is not `x` in first answer.

**Comment by Shekharrajak** [comment]
Ping @aktech @asmeurer @hargup @kshitij10496

File: sympy/solvers/decompogen.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 from sympy.polys import Poly, decompose
 
 
-def decompogen(f, symbol):
+def decompogen(*args):
```
Comment (hargup) [inline sympy/solvers/decompogen.py]: Why this change?

File: sympy/solvers/decompogen.py [hunk 2]
```diff
@@ -28,8 +29,35 @@ def decompogen(f, symbol):
...
 
     """
+    from sympy.core import Lambda
+    from sympy.geometry.util import _uniquely_named_symbol
+    from sympy.core.function import FunctionClass
+    from sympy.utilities.misc import func_name
+    from sympy.core.symbol import Symbol, Dummy
+
+    if len(args) is not 2:
+        raise ValueError('decompogen expects 2 args (function, symbol), got: %s' % len(args))
```
Comment (hargup) [inline sympy/solvers/decompogen.py]: This is rather a digression from the previous api, earlier the fact that two arguments are required was explicit, now it is implicit.

File: sympy/solvers/decompogen.py (hunk 1 above)
Comment (Shekharrajak) [inline sympy/solvers/decompogen.py]: Actually I was trying for [XFAIL](https://github.com/sympy/sympy/blob/master/sympy/solvers/tests/test_decompogen.py#L23) testcase, but I didn't find a way  for these types of cases.
I found that if we pass directly `lambda` function eg. `lambda x: x**2 + 2*x + 3` in current master branch then `function` variable contains `x**2 + 2*x + 3` so we can't extract lambda expression separately.
If we use `*args` then we can extract the `args` => `x` and `expr`=> `x**2 + 2*x + 3` using this PR changes.

File: sympy/solvers/decompogen.py [hunk 3]
```diff
@@ -28,8 +29,35 @@ def decompogen(f, symbol):
     [sin(x), sqrt(x), cos(x), x**2 + 1]
     >>> decompogen(x**4 + 2*x**3 - x - 1, x)
     [x**2 - x - 1, x**2 + x]
+    >>> decompogen(Lambda(x, sin(cos(x))), x)
```
Comment (Shekharrajak) [inline sympy/solvers/decompogen.py]: These are the testcases.

**Comment by hargup** [comment]
What is the motivation for this change, can you cite some use case?

**Comment by Shekharrajak** [comment]
@hargup , using these changes we can use `lambda` inside `decompogen`. Actually I was trying for this [XFAIL](https://github.com/Shekharrajak/sympy/blob/cf8d3c0662bffe027480a5fee770d9d3e9253d5e/sympy/solvers/tests/test_decompogen.py#L23) .

```
>>> decompogen(Lambda(x, sin(cos(x))), x)
    [sin(x), cos(x)] 
```

**Comment by hargup** [comment]
I mean why do you need to use `lambda` inside `decompogen`(not to say we shouldn't support this feature)? 

BTW what is the behavior for `decompogen(Lambda(x, sin(cos(x))), y)`? In my understanding it should be `[sin(y), cos(y)]`. Alternatively `decompgen` should take (only) Lambda functions and return Lambda functions, for example:
```python
In[1]: decompogen(Lambda(x, sin(cos(x))))
Out[1]: [Lambda(sin(x), x), Lambda(cos(x), x)]
```

As a general comment, you shouldn't add some feature unless there is sufficient motivation for that not only it takes away precious development time which can be used somewhere else, it adds lots of maintenance overhead for future.

**Comment by Shekharrajak** [comment]
>I mean why do you need to use lambda inside decompogen(not to say we shouldn't support this feature)?

When I was trying [XFAIL](https://github.com/Shekharrajak/sympy/blob/cf8d3c0662bffe027480a5fee770d9d3e9253d5e/sympy/solvers/tests/test_decompogen.py#L23)

```
In [6]: A = lambda x: x**2 + 2*x + 3

In [7]: decompogen(A(x*exp(x)), x)
Out[7]: 
⎡ 2  2⋅x        x    ⎤
⎣x ⋅ℯ    + 2⋅x⋅ℯ  + 3⎦

```
What I understood :  In `decompogen` args will be `(x*exp(x))**2 + 2*(x*exp(x)) + 3` and `x`.  So it  returns `[x**2*exp(2*x) + 2*x*exp(x) + 3]`. 

but we can try if we want to pass `lambda` directly(this PR is to share this idea).

There is no use of this PR if we don't want `lambda` as `function` in `decompogen`.

> BTW what is the behavior for decompogen(Lambda(x, sin(cos(x))), y)?

```
In [1]: decompogen(Lambda(x, sin(cos(x))), x)
Out[1]: [sin(x), cos(x)]

In [2]: decompogen(Lambda(x, sin(cos(x))), y)
Out[2]: [sin(y), cos(y), x]

In [3]: decompogen(lambda x : sin(cos(x)), x)
Out[3]: [sin(x), cos(x), x]

In [4]: decompogen(lambda x : sin(cos(x)), y)
Out[4]: [sin(y), cos(y), x]
```

I don't know whether there should be `x` in the returned list or not.

Thanks for comment.

**Comment by hargup** [comment]
> but we can try if we want to pass `lambda` directly(this PR is to share this idea).
> There is no use of this PR if we don't want `lambda` as `function` in `decompogen`.

Exactly, we don't need to support `lambda` in `decompogen` to fix the XFAIL you mentioned. To me it appears `lambda` will introduce unnecessary complications, without much benefit. So, If you have good reasons support `lambda` then we go ahead with it, otherwise we don't.

**Comment by Shekharrajak** [comment]
Okay, closing this.
